imports/importers.py

```diff
@@ -25,72 +30,150 @@
-                return shapefile.Reader(shp=shp, shx=shx, dbf=dbf)
 
-    @classmethod
-    def _import_feature(cls, fields, shape_record):
-        feature_data = dict(zip(fields, shape_record.record))
-        feature_data["geometry"] = cls._get_feature_geometry(shape_record.shape)
-
-        feature_id = feature_data.pop("id", None)
-        try:
-            feature = Feature.objects.get(id=feature_id)
-            for key, value in feature_data.items():
-                setattr(feature, key, value)
-                feature.save()
-        except Feature.DoesNotExist:
-            Feature.objects.create(**feature_data)
+    def _validate_feature_classes(self, layer):
+        """
+        Validate layer feature classes
 
-    @classmethod
-    def _get_feature_geometry(cls, shape):
-        shape_type = shape.__geo_interface__["type"].upper()
-        coordinates = "{}".format(shape.__geo_interface__["coordinates"]).replace(
-            ",", " "
+        :param layer: The layer to be validated
+        :raise ImportValidationError: if invalid feature class found
+        """
+        layer_feature_classes = set(layer.get_fields("luokkatunn"))
+        db_feature_classes = set(
+            FeatureClass.objects.filter(id__in=layer_feature_classes).values_list(
+                "id", flat=True
+            )
         )
-        geostr = "SRID={};{}{}".format(settings.SRID, shape_type, coordinates)
-        return GEOSGeometry(geostr, srid=settings.SRID)
+        missing_feature_classes = layer_feature_classes - db_feature_classes
+        if missing_feature_classes:
+            error_code = "feature_classes_not_exist"
+            raise ImportValidationError(
+                self.error_messages[error_code],
+                code=error_code,
+                params={"feature_classes": ", ".join(missing_feature_classes)},
+            )
+
+    def _import_layer(self, layer):
+        """
+        Import features from shapefile layer
+
+        :param shapefile: The path to the shapefile shp file
+        """
+        if layer.srs and layer.srs.srid != settings.SRID:
+            logger.warning(f"Invalid projection {layer.srs.srid} will be ignored.")
+
+        for feature in layer:
+            feature_data = {}
+            for shapefile_field in layer.fields:
+                model_field = SHAPEFILE_FIELD_MAPPING.get(shapefile_field)
+                if model_field:
+                    processor = FIELD_PROCESSORS.get(model_field, lambda x: x)
+                    feature_data[model_field] = processor(
+                        feature[shapefile_field].value
+                    )
+            feature_id = feature_data.pop("id")
+            geometry = feature.geom.geos
+            # The geometries in shapefile must be using the same SRID as defined in settings
+            geometry.srid = settings.SRID
+            feature_data["geometry"] = feature.geom.geos
```

Comment:
This is somewhat odd. Why first name it "geometry" and then refer to it again as "feature.geom.geos"?

Reply:
good catch there, it should be using geometry